xds/src/main/java/io/grpc/xds/EnvoyProtoData.java

```diff
@@ -1113,6 +1138,270 @@ public String toString() {
+          && Objects.equals(delayNanos, that.delayNanos);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(delayNanos, ratePerMillion);
+    }
+
+    @Override
+    public String toString() {
+      ToStringHelper toStringHelper = MoreObjects.toStringHelper(this)
+          .add("ratePerMillion", ratePerMillion);
+      if (delayNanos != null) {
+        toStringHelper.add("type", "fixed delay");
+        toStringHelper.add("delayNanos", delayNanos);
+      } else {
+        toStringHelper.add("type", "header delay");
+      }
+      return toStringHelper.toString();
+    }
+
+    private static FaultDelay fromEnvoyProtoFaultDelay(
+        io.envoyproxy.envoy.extensions.filters.common.fault.v3.FaultDelay faultDelay) {
+      int rate = getRatePerMillion(faultDelay.getPercentage());
+      if (faultDelay.hasHeaderDelay()) {
+        return new FaultDelay(null, rate);
+      }
+      long delay = Durations.toNanos(faultDelay.getFixedDelay());
+      return new FaultDelay(delay, rate);
+    }
+  }
+
+  /**
+   * See corresponding Envoy proto message {@link
+   * io.envoyproxy.envoy.extensions.filters.http.fault.v3.FaultAbort}.
+   */
+  static final class FaultAbort {
+    @Nullable
+    final Status status;
+    final int ratePerMillion;
+
+    FaultAbort(@Nullable Status status, int ratePerMillion) {
+      this.status = status;
+      this.ratePerMillion = ratePerMillion;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      FaultAbort that = (FaultAbort) o;
+      return ratePerMillion == that.ratePerMillion
+          && Objects.equals(status, that.status);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(status, ratePerMillion);
+    }
+
+    @Override
+    public String toString() {
+      ToStringHelper toStringHelper = MoreObjects.toStringHelper(this)
+          .add("ratePerMillion", ratePerMillion);
+      if (status != null) {
+        toStringHelper.add("type", "fixed status");
+        toStringHelper.add("status", status);
+      } else {
+        toStringHelper.add("type", "header abort");
+      }
+      return toStringHelper.toString();
+    }
+
+    private static StructOrError<FaultAbort> fromEnvoyProtoFaultAbort(
+        io.envoyproxy.envoy.extensions.filters.http.fault.v3.FaultAbort faultAbort) {
+      int rate = getRatePerMillion(faultAbort.getPercentage());
+      switch (faultAbort.getErrorTypeCase()) {
+        case HEADER_ABORT:
+          return StructOrError.fromStruct(new FaultAbort(null, rate));
+        case HTTP_STATUS:
+          Status status = convertHttpStatus(faultAbort.getHttpStatus());
+          return StructOrError.fromStruct(new FaultAbort(status, rate));
+        case GRPC_STATUS:
+          Code code;
+          try {
+            code = Code.values()[faultAbort.getGrpcStatus()];
```

Comment:
Should use `Status.fromCodeValue(...)` and illegal gRPC code value should be handled in the gRPC's way.